proposals/js/JsName.md

```diff
@@ -0,0 +1,197 @@
+
+*__Question:__ how it should work for overrides?*
+* allow only on final
+* prohibit on overrides and reuse names from overridden declaration
+    * how, then, resolve conflicts?
+
++ `PROPERTY_GETTER` and `PROPERTY_SETTER`
+<br/>
+Can be useful for native declarations to provide better api in Kotlin.
+Additionally it can be used to force using functions for accessors to avoid problems with some minifiers
+(e.g. closure-compiler) which treat access to properties as side effect free.
+<br/>
+    *__Frontend:__ the annotation can not be applied to only one of accessors.*
+<br/>
+    *__Frontend:__ the annotation can not be simultaneously applied to the property and its accessors.*
+<br/>
+    *__Question:__ should it be prohibited for non-native declarations?*
+<br/>
+    *__Backend:__ properties whose accessors has this annotation no longer are treated as JS properties,
+    instead they are interpreted as bunch of accessors (like in JVM), so backend:*
+    - *doesn't generate JS property for them;*
+    - *translates accessors as functions using name from the annotation;*
+    - *translates an accessing to such property as call to accessor function using name from the annotation.*
+
+
+## How it should work for overrides?
+
+Actually, the problems actual even we prohibit using `JsName` on overrides,
+because the compiler doesn't mangle the name of native declarations,
+so we can simply get name clash when to inherit from two natives.
+
+And doesn't make sense to prohibit to use this annotation on non-final declarations.
+
+So propose to **prohibit using JsName on overrides**, but **report about name clashes** on non-native declarations.
+
+
+**Case 1: declarations with different names from different parents with the same requested name (by JsName)**
+    _(map many names to the one)_
+
+
+_Note: A, B, C can be native._
+
+Example 1:
+```kotlin
+interface A {
+    @JsName("boo")
+    fun bar()
+}
+
+
+interface B {
+    @JsName("boo")
+    fun baz()
+}
+
+class C : A, B {
+    override fun bar() {}
+    override fun baz() {}
+}
+```
+
+Example 2:
+```kotlin
+interface A {
+    @JsName("boo2")
+    fun bar2() {}
+}
+
+
+interface B {
+    @JsName("boo2")
+    fun baz2() {}
+}
+
+class C : A, B
+```
+
+Proposed solution: allow name clashes on native declarations and prohibit on non-native declarations.
+
+**Case 2: declarations with the same name from different parents with different requested name (by JsName):**
+    _(map the one name to many)_
+
+_Note: A, B, C can be native._
+
+Example 1:
+```kotlin
+interface A {
+    @JsName("fooA")
+    fun foo()
+}
+
+
+interface B {
+    @JsName("fooB")
+    fun foo()
+}
+
+class C : A, B {
+    override fun foo() {}
+}
+```
+
+Example 2:
+```kotlin
+interface A {
+    @JsName("fooA")
+    fun foo2() {}
+}
+
+
+interface B {
+    @JsName("fooB")
+    fun foo2()
+}
+
+class C : A, B
+```
+
+Proposed solution: prohibit to have more than one name candidate for all declarations including native,
```

Comment:
In general, there is case when we get different name candidates in the similar case:

```
interface A<T> {
    fun foo(value: T)
}
open class B {
    open fun foo(value: Int)
}
class C : B, A<Int> {
}
```

in this case we generate bridges and don't report error. So we should prohibit only certain cases. In my prototype I report error only when one of the candidates has `@JsName` on it.